kernel/fatal.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2019 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <kernel_internal.h>
+#include <sys/printk.h>
+#include <arch/cpu.h>
+#include <logging/log_ctrl.h>
+#include <logging/log.h>
+#include <fatal.h>
+
+/* LCOV_EXCL_START */
+FUNC_NORETURN __weak void z_arch_system_halt(unsigned int reason)
+{
+	ARG_UNUSED(reason);
+
+	/* TODO: What's the best way to totally halt the system if SMP
+	 * is enabled?
+	 */
+
+	(void)z_arch_irq_lock();
+	for (;;) {
+		k_cpu_idle();
+	}
+}
+/* LCOV_EXCL_STOP */
+
+/* LCOV_EXCL_START */
+__weak void k_sys_fatal_error_handler(unsigned int reason,
+				      const z_arch_esf_t *esf)
+{
+	ARG_UNUSED(esf);
+
+	z_fatal_print("Halting system");
+	z_arch_system_halt(reason);
+	CODE_UNREACHABLE;
+}
+/* LCOV_EXCL_STOP */
+
+#if defined(CONFIG_PRINTK) || defined(CONFIG_LOG)
+void z_fatal_print(const char *fmt, ...)
+{
+	va_list ap;
+
+	LOG_PANIC();
+
+	va_start(ap, fmt);
+#ifdef CONFIG_LOG
+	struct log_msg_ids src_level = {
+		.level = LOG_LEVEL_ERR
+	};
+
+	log_generic(src_level, fmt, ap);
+#else
+	printk("FATAL: ");
+	vprintk(fmt, ap);
+	printk("\n");
+#endif
+	va_end(ap);
+}
+
+static void exc_print(char *name)
+{
+	/* sanitycheck looks for the "ZEPHYR FATAL ERROR" string, don't
+	 * change it without also updating sanitycheck
+	 */
+	z_fatal_print(">>> ZEPHYR FATAL ERROR: %s", name);
+}
+#endif
+
+void z_fatal_error(unsigned int reason, const z_arch_esf_t *esf)
+{
+#if defined(CONFIG_PRINTK) || defined(CONFIG_LOG)
+	struct k_thread *thread = k_current_get();
+	const char *thread_name = k_thread_name_get(thread);
+
+	if (thread_name == NULL || thread_name[0] == '\0') {
+		thread_name = "unknown";
+	}
+
+	switch (reason) {
```

Comment:
imo, it would be cleaner to extract it to function `reason_to_str()` and do something like `z_fatal_print(">>> ZEPHYR FATAL ERROR: %s (reason %d)", reason_to_str(reason), reason);`. it will be also less rom used since there will be only one call to printing instead of multiple.

Reply:
Done